spinalcordtoolbox/reports/qc2.py

```diff
@@ -150,6 +151,777 @@ def create_qc_entry(
+
+        # Take into account the size of the cropping rectangle.
+        offsets[:, 1:3] -= (np.array(shape) - 1) / 2
+
+        return SlicingSpec(
+            affine=self.affine,
+            offsets=dict(zip(self.offsets.keys(), offsets)),
+            shape=shape,  # new shape
+            axis_labels=self.axis_labels,
+        )
+
+    def center_columns(self: 'SlicingSpec', seg: Image) -> 'SlicingSpec':
+        """
+        A re-centered and cropped slicing spec, meant for sagittal views.
+
+        Each slice is individually centered in the last axis around the center
+        of mass of the given segmentation. It is also cropped in the last axis
+        to the larger of:
+        - 110% of the maximum width of the segmentation, or
+        - 50% of the total image.
+
+        Any slices that are more than 2 slices away from the segmentation are
+        also dropped.
+        """
+        # Check which slices contain the segmentation, and compute their width.
+        slices_seg = self.get_slices(seg, order=1)
+        width = {}
+        for i, slice_seg in enumerate(slices_seg.values()):
+            [columns] = slice_seg.any(axis=0).nonzero()
+            if columns.size != 0:
+                width[i] = int(columns.max()) - int(columns.min()) + 1
+        if not width:
+            raise ValueError("The mask image is empty. Cannot crop using an empty mask. Check the input (e.g. '-qc-seg').")
+
+        # Compute the target slice width.
+        shape = (
+            self.shape[0],  # the full image height
+            math.ceil(max(
+                1.1 * max(width.values()),  # 110% of the segmentation width
+                0.5 * self.shape[1],  # 50% of the full image width
+            )),
+        )
+
+        # Drop slices that are far from the segmentation.
+        slice_min = max(0, min(width.keys()) - 2)
+        slice_max = min(len(slices_seg) - 1, max(width.keys()) + 2)
+        offsets_dict = dict(list(self.offsets.items())[slice_min:slice_max+1])
+
+        # Re-center each remaining slice.
+        for slice_label, offset in offsets_dict.items():
+            offset[2] += ndimage.center_of_mass(slices_seg[slice_label])[1]
+
+        # Take into account the cropping width.
+        offsets_array = np.array(list(offsets_dict.values()))
+        inf_nan_fill(offsets_array[:, 2])
+        offsets_array[:, 2] -= (shape[1] - 1) / 2
+        offsets_dict = dict(zip(offsets_dict.keys(), offsets_array))
+
+        return SlicingSpec(
+            affine=self.affine,
+            offsets=offsets_dict,  # new offsets
+            shape=shape,  # new shape
+            axis_labels=self.axis_labels,
+        )
+
+    def center_lines(self: 'SlicingSpec', seg: Image) -> 'SlicingSpec':
+        """
+        A re-centered and cropped slicing spec, meant to transform a full axial
+        slicing spec into a collection of single-line slices which can be
+        arranged into a single wavy mid-sagittal image.
+        """
```

Comment:
Nit, but the name and description come across to me as a little difficult to grok. (Specifically, the notion of `center_lines` != `centerlines`, and the terms "recentered and cropped" as well as "wavy".)

I absolutely get what you mean by the name + terms, but I wonder if something like this might be a little more intuitive?

```suggestion
    def composite_sagittal(self: 'SlicingSpec', seg: Image) -> 'SlicingSpec':
        """
        Construct a composite 2D sagittal image out of individual rows 
        (one from each axial slice), similarly to `sct_flatten_sagittal`.

        (For each axial slice, the center of mass is used to identify the 
        midsagittal row. The individual rows can then be stitched together 
        to build a composite midsagittal 2D image.)
        """
```